### PR TITLE
ITV2: disallow duplicate dynamic templates

### PR DESCRIPTION
**Pull request by andreidan**
Dynamic templates can contain multiple templates with the same name. This
commit changes the way V2 index templates mappings are resolved to add
deduplication of dynamic templates when resolving the mappings, by having
the last dynamic templates override the ones defined with the same name
earlier in the `dynamic_templates` array.

Relates to #53101
Relates to #53326
Relates to #28988

### CODE REVIEW

**Comment by dakrone** [comment]
@andreidan can you add a test that component templates merge their dynamic templates correctly? I think because we merge the XContent maps prior to the dedupe it is possible to end up with an invalid dynamic template if you have two component templates that each define a dynamic template with the same name but different configurations that, when merged, make the template invalid.

**Comment by andreidan** [comment]
@dakrone thanks for the suggestion on the component templates merging. Great shout on testing that as you were right and we are merging the Maps so, even though we dedup the key names (ie. there will only be one `dynamic_template` with a particular name) the bodies are merged.

**Comment by andreidan** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java [hunk 1]
```diff
@@ -576,6 +577,18 @@ private ClusterState applyCreateIndexRequestWithExistingMetadata(final ClusterSt
...
+                    "dynamic_templates");
+                List<Map<String, Object>> previouslySeenDynamicTemplates =
+                    (List<Map<String, Object>>) nonProperties.get("dynamic_templates");
+                // we want "later" defined dynamic templates to override the previously seen ones with the same name
+                List<Map<String, Object>> filteredDefaultDynamicTemplates =
+                    removeAll(previouslySeenDynamicTemplates, innerTemplateDynamicTemplates);
+
+                if (filteredDefaultDynamicTemplates != previouslySeenDynamicTemplates) {
+                    nonProperties.put("dynamic_templates", filteredDefaultDynamicTemplates);
+                }
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java:590]: I think we can encapsulate this logic into a method to make it a bit clearer what's going on, perhaps something like:

```
nonProperties = removeDuplicatedDynamicTemplates(nonProperties, innerTemplateNonProperties);
```

Then the same method can be used below (where these lines are almost identical except for the maps passed in).

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java [hunk 2]
```diff
@@ -598,11 +623,78 @@ private ClusterState applyCreateIndexRequestWithExistingMetadata(final ClusterSt
...
+     *
+     * This expects both list values to be Maps of size one and the "contains" operation that will determine if a value
+     * from the second list is present in the first list (and be removed from the first list) is based on key name.
+     *
+     * eg.
+     *      removeAll([ {"key1" : {}}, {"key2" : {}} ], [ {"key1" : {}}, {"key3" : {}} ])
+     * Returns:
+     *     [ {"key2" : {}} ]
+     */
+    private static List<Map<String, Object>> removeAll(List<Map<String, Object>> first, List<Map<String, Object>> second) {
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java:644]: Maybe call this `removeOverlapping` to make it a bit clearer what it's doing?

**Review by dakrone — CHANGES_REQUESTED** [review]
I left a couple of comments similar to what we talked about regarding naming, thanks for doing this Andrei!

**Review by dakrone — APPROVED** [review]
LGTM, thanks for working on this Andrei!

**Pull request merged**